File: universal-ai-artifacts/utils/rate_limit.js
export async function rateLimitedFetch(fetchFn, url, options = {}, {
  maxRetries = 4,
  baseDelayMs = 500,
  backoffFactor = 2,
  maxDelayMs = 8000
} = {}) {
  let attempt = 0;
  const originalBody = options.body;

  while (true) {
    try {
      // Re-clone simple bodies for retries
      if (attempt > 0 && typeof originalBody === 'string') {
        options.body = originalBody;
      }

      const response = await fetchFn(url, options);
      if (response.status !== 429 && response.status < 500) {
        return response; // success or client error
      }

      // 429 or 5xx -> retry
      if (attempt >= maxRetries) {
        return response; // give back last response
      }

      const retryAfter = parseInt(response.headers.get('Retry-After') || '0', 10);
      const delayFromHeader = Number.isFinite(retryAfter) && retryAfter > 0 ? retryAfter * 1000 : 0;
      const backoff = Math.min(baseDelayMs * Math.pow(backoffFactor, attempt), maxDelayMs);
      const jitter = Math.floor(Math.random() * 250);
      const delayMs = Math.max(delayFromHeader, backoff) + jitter;

      await new Promise((r) => setTimeout(r, delayMs));
      attempt += 1;
      continue;
    } catch (err) {
      if (attempt >= maxRetries) throw err;
      const backoff = Math.min(baseDelayMs * Math.pow(backoffFactor, attempt), maxDelayMs);
      const jitter = Math.floor(Math.random() * 250);
      await new Promise((r) => setTimeout(r, backoff + jitter));
      attempt += 1;
    }
  }
}

File: universal-ai-artifacts/utils/rate_limit.py
import time
import random
from typing import Callable, Dict, Any


def request_with_retries(sender: Callable[..., Any], max_retries: int = 4, base_delay: float = 0.5, backoff: float = 2.0, max_delay: float = 8.0, **kwargs) -> Any:
    """
    rate-limit-safe request wrapper. Expects `sender` to be a callable like
    `session.request(method, url, **kwargs)` returning a `requests.Response`.
    Retries on HTTP 429 and 5xx. Honors Retry-After header when present.
    """
    attempt = 0
    while True:
        try:
            resp = sender(**kwargs)
            if resp.status_code != 429 and resp.status_code < 500:
                return resp
            if attempt >= max_retries:
                return resp
            retry_after = resp.headers.get('Retry-After')
            delay_from_header = 0.0
            if retry_after is not None:
                try:
                    delay_from_header = float(retry_after)
                except ValueError:
                    delay_from_header = 0.0
            backoff_delay = min(base_delay * (backoff ** attempt), max_delay)
            delay = max(delay_from_header, backoff_delay) + random.uniform(0, 0.25)
            time.sleep(delay)
            attempt += 1
        except Exception:
            if attempt >= max_retries:
                raise
            backoff_delay = min(base_delay * (backoff ** attempt), max_delay)
            time.sleep(backoff_delay + random.uniform(0, 0.25))
            attempt += 1